enclave/src/main/java/org/hyperledger/besu/enclave/Enclave.java

```diff
@@ -26,101 +28,117 @@
+    final CompletableFuture<Boolean> result = new CompletableFuture<>();
+    final HttpClientRequest request = client.request(HttpMethod.GET, "/upcheck").setTimeout(1000);
+    request.end(asyncResult -> result.complete(asyncResult.succeeded()));
 
-    try (Response response = client.newCall(request).execute()) {
-      return response.isSuccessful();
-    } catch (IOException e) {
-      LOG.error("Enclave failed to execute upcheck");
-      throw new IOException("Failed to perform upcheck", e);
-    }
+    return result.get();
   }
 
   public SendResponse send(final SendRequest content) {
-    return executePost(buildPostRequest(JSON, content, "/send"), SendResponse.class);
+    return postRequest(JSON, content, "/send", SendResponse.class);
   }
 
   public ReceiveResponse receive(final ReceiveRequest content) {
-    return executePost(buildPostRequest(ORION, content, "/receive"), ReceiveResponse.class);
+    return postRequest(ORION, content, "/receive", ReceiveResponse.class);
   }
 
   public PrivacyGroup createPrivacyGroup(final CreatePrivacyGroupRequest content) {
-    return executePost(buildPostRequest(JSON, content, "/createPrivacyGroup"), PrivacyGroup.class);
+    return postRequest(JSON, content, "/createPrivacyGroup", PrivacyGroup.class);
   }
 
   public String deletePrivacyGroup(final DeletePrivacyGroupRequest content) {
-    return executePost(buildPostRequest(JSON, content, "/deletePrivacyGroup"), String.class);
+    return postRequest(JSON, content, "/deletePrivacyGroup", String.class);
   }
 
   public PrivacyGroup[] findPrivacyGroup(final FindPrivacyGroupRequest content) {
-    Request request = buildPostRequest(JSON, content, "/findPrivacyGroup");
-    return executePost(request, PrivacyGroup[].class);
+    return postRequest(JSON, content, "/findPrivacyGroup", PrivacyGroup[].class);
   }
 
-  private Request buildPostRequest(
-      final MediaType mediaType, final Object content, final String endpoint) {
-    final RequestBody body;
+  private <T> T postRequest(
+      final AsciiString mediaType,
+      final Object content,
+      final String endpoint,
+      final Class<T> responseType) {
+
     try {
-      body = RequestBody.create(mediaType, objectMapper.writeValueAsString(content));
+      final CompletableFuture<T> result = new CompletableFuture<>();
+      final HttpClientRequest request = client.request(HttpMethod.POST, endpoint);
+      request.handler(response -> handleResponse(response, responseType, result));
+      request.putHeader("content-Type", mediaType);
+      request.exceptionHandler(result::completeExceptionally);
+      request.setChunked(false);
+      request.end(objectMapper.writeValueAsString(content));
+      return result.get();
     } catch (final JsonProcessingException e) {
       throw new EnclaveException("Failed to serialise request to json body.");
+    } catch (final ExecutionException e) {
+      if (e.getCause() instanceof EnclaveException) {
```

Comment:
why do you need to check the cause is a EnclaveException?

Reply:
getCause() returns a Throwable, not an exception, so you can't (easily) just rethrow e.getCause() - this way, we throw the underlying exception if it was understood by the thrower - otherwise, we throw a sort of "Unknown Enclave problem".